### ISSUE TITLE
[Question]:MTP原理

### ISSUE THREAD

**Issue by pennypengpm** [PaddlePaddle/PaddleNLP#10258]
![Image](https://github.com/user-attachments/assets/7537e77b-0698-491a-a50f-7865f1d7390e)

您好，我尝试跑deepseekR1-MTP，对于上述的原理图没有看懂，请问具体是怎么实现的呢

**Comment by freeliuzc** [comment]
1. 上图 batch_extension 的具体做法可以参考 vllm 的文档 https://docs.google.com/document/d/1T-JaS2T1NRfdP51qzqpyakoCXxSXTtORppiwaj5asxA/edit?tab=t.0
2. 下图我们的实现，可以理解为在 decoder 阶段，支持 num_token > 1, 即 MTP 模型产出任意个 token（num_draft_token），在 base model 验证时都可以直接推理，一次性产出 (num_draft_token+1) 个 logits 结果

**Comment by pennypengpm** [comment]
感谢，再请问一下2具体是指mla计算是指这里么， https://github.com/PaddlePaddle/PaddleNLP/blob/develop/csrc/gpu/mla_attn/mla_hopper.cuh#L68

**Comment by pennypengpm** [comment]
@freeliuzc 

![Image](https://github.com/user-attachments/assets/e1585760-b2fd-49c2-8c63-c8434424e937)
这里说batch_size=128, num_draft_token=2 时，batch 规模大小仍为 128
![Image](https://github.com/user-attachments/assets/fa6f3c19-7d30-4639-829f-159caa7d80df)
Q这里的token_num还是batch_size*num_draft_token，还是没太理解怎么减少了计算量

**Comment by freeliuzc** [comment]
1. 上面的链接这个图，蓝色部分是现在很多开源实现方式的时延，红色的是我们实现方式的时延。建议先仔细阅读下这篇文档~
2.  batch_extension 是指 batch 要膨胀 draft_token 倍，但支持 num_token > 1 后，就无需膨胀了
<img width="1084" alt="Image" src="https://github.com/user-attachments/assets/51a76f6e-3e63-4fb1-800f-65c737f311e0" />

> 1. 上图 batch_extension 的具体做法可以参考 vllm 的文档 https://docs.google.com/document/d/1T-JaS2T1NRfdP51qzqpyakoCXxSXTtORppiwaj5asxA/edit?tab=t.0
> 2. 下图我们的实现，可以理解为在 decoder 阶段，支持 num_token > 1, 即 MTP 模型产出任意个 token（num_draft_token），在 base model 验证时都可以直接推理，一次性产出 (num_draft_token+1) 个 logits 结果
